t3f/tensor_train_base.py

```diff
@@ -0,0 +1,159 @@
+import numpy as np
+import tensorflow as tf
+
+
+# TODO: check the methods of _TensorLike
+class TensorTrainBase(object):
+  """An abstract class that represents a collection of Tensor Train cores.
+  ```
+  @@__init__
+  @@get_raw_shape
+  @@get_shape
+  @@tt_cores
+  @@dtype
+  @@name
+  @@graph
+  @@ndims
+  @@get_tt_ranks
+  @@is_tt_matrix
+  @@is_variable
+  @@eval
+  """
+
+  def __init__(self, tt_cores):
+    """Creates a `TensorTrainBase`."""
+    pass
+
+  def get_raw_shape(self):
+    """Get tuple of `TensorShapes` representing the shapes of the underlying TT-tensor.
+
+    Tuple contains one `TensorShape` for TT-tensor and 2 `TensorShapes` for
+    TT-matrix
+
+    Returns:
+      A tuple of `TensorShape` objects.
+    """
+    return self._raw_shape
+
+  def get_shape(self):
+    """Get the `TensorShape` representing the shape of the dense tensor.
+
+    Returns:
+      A `TensorShape` object.
+    """
+    raw_shape = self.get_raw_shape()
+    if self.is_tt_matrix():
+      # TODO: as list is not available if shape is partly known.
+      m = np.prod(raw_shape[0].as_list())
+      n = np.prod(raw_shape[1].as_list())
+      return tf.TensorShape((m, n))
+    else:
+      return raw_shape[0]
+
+  @property
+  def tt_cores(self):
+    """A tuple of TT-cores."""
+    return self._tt_cores
+
+  @property
+  def dtype(self):
+    """The `DType` of elements in this tensor."""
+    # TODO: where is this created?
+    return self.tt_cores[0].dtype
+
+  @property
+  def name(self):
+    """The name of the TensorTrain.
+
+    Returns:
+      String, the scope in which the TT-cores are defined.
+    """
+    core_name = self.tt_cores[0].name
+    idx = core_name.rfind('/')
+    return core_name[:idx]
+
+  @property
+  def graph(self):
+    """The `Graph` that contains the tt_cores tensors."""
+    # TODO: check in init that the other cores are from the same graph.
+    return self.tt_cores[0].graph
+
+  def __str__(self):
+    """A string describing the TensorTrain object, its TT-rank and shape."""
+    return NotImplementedError
+
+  def ndims(self):
+    """Get the number of dimensions of the underlying TT-tensor.
+
+    Returns:
+      A number.
+    """
+    return len(self.tt_cores)
+
+  def get_tt_ranks(self):
+    """Get the TT-ranks in an array of size `num_dims`+1.
+
+    The first and the last TT-rank are guarantied to be 1.
+
+    Returns:
+      TensorShape of size `num_dims`+1.
+    """
+    return self._tt_ranks
+
+  def is_tt_matrix(self):
+    """Returns True if the TensorTrain object represents a TT-matrix."""
+    return len(self.get_raw_shape()) == 2
```

Comment:
I think in this particular case it's OK (but definitely not great) since there should be exactly one occurrence of "== 2" (not sure where did you saw the others).